webapp/components/Map/KmlUploader.js

```diff
@@ -0,0 +1,145 @@
+import './KmlUploader.scss'
+
+import React, { useEffect, useState } from 'react'
+import { useMap } from 'react-leaflet'
+
+import * as User from '@core/user/user'
+import { useUser } from '@webapp/store/user'
+import { useI18n } from '@webapp/store/system'
+
+import L from 'leaflet'
+
+require('./L.KML')
+
+import JSZip from 'jszip'
+import shp from 'shpjs'
+
+export const KmlUploader = () => {
+  const map = useMap()
+  const jszip = new JSZip()
+
+  const i18n = useI18n()
+  const user = useUser()
+
+  const [selectedFile, setSelectedFile] = useState()
+  const [layers, setLayers] = useState([])
+  const [opacity, setOpacity] = useState(50)
+
+  useEffect(() => {
+    if (selectedFile) {
+      if (selectedFile.name.endsWith('.kmz')) {
+        processKMZFile(selectedFile)
+      } else if (selectedFile.name.endsWith('.kml')) {
+        processKMLFile(selectedFile)
+      } else if (selectedFile.name.endsWith('.zip')) {
+        processShapeFile(selectedFile)
+      }
+    }
+  }, [selectedFile])
+
+  const processShapeFile = (file) => {
+    const geo = L.geoJson(
+      { features: [] },
+      {
+        onEachFeature: function popUp(f, l) {
+          if (f.properties) {
+            const out = Object.entries(f.properties).map((key) => key + ': ' + f.properties[key])
+            l.bindPopup(out.join('<br />'))
+          }
+        },
+      }
+    ).addTo(map)
+    setLayers((old) => [...old, geo])
+    const reader = new FileReader()
+    reader.onload = async (e) => {
+      const text = e.target.result
+      shp(text).then(function (data) {
+        geo.addData(data)
+      })
+    }
+    reader.readAsArrayBuffer(file)
+  }
+
+  const processKMLFile = (file) => {
+    const reader = new FileReader()
+    reader.onload = async (e) => {
+      const text = e.target.result
+      addKMLLayers([text])
+    }
+    reader.readAsText(file)
+  }
+
+  const processKMZFile = async (file) => {
+    const kmlList = []
+    let promises = []
+    await jszip.loadAsync(file).then(() => {
```

Comment:
why do you have an "await" and then a call to "then"?
you should be able to do just: 
await jszip.loadAsync(file)
jszip.forEach(...
await Promise.all(promises)
addKMLLayers(kmlList)